### ISSUE TITLE
Joining rooms over federation can be very slow (SYN-293)

### ISSUE THREAD

**Issue by matrixbot** [matrix-org/synapse#1211]
__Edit:__ Currently tracking related work in the milestones:
 * https://github.com/matrix-org/synapse/milestone/6
 * https://github.com/matrix-org/synapse/milestone/8
 * https://github.com/matrix-org/synapse/milestone/10

For an overview of the technical changes, see MSC3902.

---

__Original Post:__

I have Synapse/0.7.1-r2 installed from pip/github. 
I am running nginx on a public server in front of it:
 - https://matrix.ytnoc.net:443/ -> proxied to http 8009
 - https://matrix.ytnoc.net:8448/ -> proxied to https 8448
I have copied my key and crt in use on the nginx server down to my homeserver install and restarted (had to do this before federation worked).

I can initiate 1:1 chats or join rooms on matrix.org from matrix.ytnoc.net. If I join an active room like #matrix-dev:matrix.org, it takes well over a minute to synchronize. I only had one instance where I saw all the user avatars and the channel name show up.  The rest of the time, the channel name on the left is replace my my local id (@&#8203;kb3dfz:matrix.ytnoc.net).  See screenshot.

#test:matrix.org seems to be working just fine with federation.

I have attached homeserver.log

(Imported from https://matrix.org/jira/browse/SYN-293)

(Reported by John Hogemiller)

#### Attachments:
https://matrix.org/jira/secure/attachment/10115/icognito-1.png
https://matrix.org/jira/secure/attachment/10118/noalias.png
https://matrix.org/jira/secure/attachment/10116/redacted_homeserver.log
https://matrix.org/jira/secure/attachment/10113/Screen+Shot+2015-02-28+at+11.24.53+AM.png
https://matrix.org/jira/secure/attachment/10117/usernames+only.png

**Comment by richvdh** [comment]
Having just done a join from a test server to matrix HQ:
```
The `join` eventually completed with:

```
2018-04-18 13:55:23,582 - synapse.access.https.8447 - 93 - INFO - POST-726- None - 8447 - {@richvdh:matrixtest.sw1v.org} Processed request: 467006ms (53781ms, 4376ms) (12344977ms/16508ms/4567) 65B 200 "POST /_matrix/client/r0/join/%23matrix%3Amatrix.org? HTTP/1.1" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36"
```

- that's 12345 seconds (3.5 hours) waiting for db connections. given the whole request completed in "only" 467 seconds, presumably it is doing a bunch of that waiting in parallel.

I suspect we're handling each server separately and in parallel from the point of view of fetching server keys and persisting said keys to the db. We should apply more intelligence here.

**Comment by richvdh** [comment]
Also related: #3120

**Comment by richvdh** [comment]
Some things which are related here:
* #2418
* #3818
* #2956

**Comment by zpxp** [comment]
I still get this issue with the latest official synapse docker image and using postgres db

**Comment by Janhouse** [comment]
Having similar experience without being able to join any larger rooms (~1000 users) over federated connection.
Should it actually be possible?

**Comment by FSG-Cat** [comment]
Yes its completely possible. A lot of the matrix devs and enthusiaists run their own homeservers and quite a few are in rooms like HQ and HQ is currently at 25k members so ye joining 1000+ user rooms over Fed is fully possible and works.

**Comment by nickian** [comment]
I migrated to PostgreSQL and even upped my server to dedicated, 8GB memory, 4 CPUs. No difference whatsoever. Trying to join #matrix:matrix.org:

![image](https://user-images.githubusercontent.com/590678/139600122-d9e56543-fe25-431c-b399-3da03a9c17e4.png)

**Comment by nickian** [comment]
It randomly joined finally when I logged back in a half hour later, despite the error.

**Comment by BBaoVanC** [comment]
> It randomly joined finally when I logged back in a half hour later, despite the error.

Yeah it's a bit of a UI problem in Element. It shouldn't say there was an error, because the server is probably still trying to join anyways.

**Comment by enthus1ast** [comment]
Since three days i try to join a big room ~5000 members.
Everytime i get:

![image](https://user-images.githubusercontent.com/13794470/141313481-8fec11a0-3e65-4fff-9f85-ed6bf1fe15b0.png)

**Comment by squahtx** [comment]
@enthus1ast That may be #11135, which is fixed in the coming 1.47.0 release.

Could you check the Synapse logs for the full error message? If the error contains `TypeError: Object of type Decimal is not JSON serializable` then it's definitely #11135 and upgrading to a 1.47 rc ought to resolve your issue. Otherwise please file a new issue, including your Synapse version and the full error from the logs.

**Comment by enthus1ast** [comment]
@squahtx yes exactly this issue. Thanks for clarification.

**Comment by Axeltherabbit** [comment]
same issue, so frustrating and it lags the server a lot too

**Comment by zippyy** [comment]
I am having the same issue. I am running a vm in proxmox with 2 vcpu and 6gb ecc ram. I can join some rooms but not others and ones i can join do take a long time to join. I have this issue both with me deployment via the ansible deploy playbook as well as manual deployment via docker.

**Comment by amstan** [comment]
Yeah... this business of taking at least 5 min, maybe 20, to join for any random project's channel (which used to be an instant join with irc) is getting old.

Element reports 502 errors and blocks the UI with a model window until then, sometimes i have to try the join multiple times (since the first one doesn't take).

I have a 32GB machine, with postgresql.

7 year old bug, sigh.

**Comment by deepbluev7** [comment]
If you people missed it, there is work being done on this: https://matrix.org/blog/2022/01/28/this-week-in-matrix-2022-01-28#synapse-website

**Comment by tucnak** [comment]
@richvdh First of all, thank you so much for putting in the work to battle this awful issue that plagued Matrix since forever! Could you please be kind to provide some simple up-to-date note/overview on the ongoing efforts so those of us eagerly awaiting, and possibly looking to contribute— could understand this better? I, for one, am very interested in resolving this, and I'm sure many home server maintainers, too would be happy to chip in, but the codebase and the intrinsics seem rather intimidating given there's limited granularity in the issuing for SYN-293.

**Comment by richvdh** [comment]
I've put some notes on progress up at https://hackmd.io/@richvdh/SkhcKnjKY.

**Comment by neoromantique** [comment]
+1 This is very annoying and breaks new user UX immensely.

Even if it is not possible to speed it up, at least providing some feedback that it is happening in the background would be great

**Comment by tucnak** [comment]
@richvdh 404 Not Found!

**Comment by mind-overflow** [comment]
Hey, is there any progress on this? Joining big rooms still takes an annoyingly long time on homeservers.

**Comment by deepbluev7** [comment]
Please stop spamming this issue with "+1"s or "this affects me too". It's not helpful. If it affects you, add a thumbs-up to the issue.

As about progress, there is no writeup for it, but you can follow along by reading the changelogs and looking for lines containing "faster" and "join". I.e. https://github.com/matrix-org/synapse/releases/tag/v1.60.0 mentions 2 things done as preparatory work.

You can also take a look at https://github.com/matrix-org/synapse/milestone/6 which lists some of the work done on this. But some pull requests like https://github.com/matrix-org/synapse/pull/12877 are also done as part of the faster joins work, but not explicitly listed.

So in general, you can find the current progress by using the search on this repository, since most PRs and issues mention "faster joins". As you can see it is actively being worked on, but also quite technical, so for most people the updates or explanations won't make sense.

Also, don't ask when it will be done. It will take longer every time you ask :3

**Comment by richvdh** [comment]
> @richvdh 404 Not Found!

Sorry about that; I've updated the link to my notepad. However, the milestones (https://github.com/matrix-org/synapse/milestone/6, https://github.com/matrix-org/synapse/milestone/8) are probably a better place to follow along.

**Comment by gorbehnare** [comment]
Hello, I'm just trying to figure out if this is the problem I am seeing, or if something else is broken. I set up my first synapse server last week, and within two days I had everything working (including VoIP and video calls). By this point, I had already joined the public synapse-admins room and was able to chat, ask and answer questions there. As of this week, I keep getting "no known server" and "failed to join" errors after very long wait times when trying to join (seemingly) any public room. 
In frustration, I have scrapped the VM, and installed a new VM with more RAM, a new domain, postgresSQL, etc... all from scratch, and I am still facing this. What are the test cases for this? how can I find out if the issue is with my homeserver or just a general matrix issue here?

**Comment by richvdh** [comment]
> how can I find out if the issue is with my homeserver or just a general matrix issue here?

Find the `/join` request from your client to your server in your homeserver logs, then look at the other lines logged for that request (see https://matrix-org.github.io/synapse/latest/usage/administration/admin_faq.html#how-can-i-find-the-lines-corresponding-to-a-given-http-request-in-my-homeserver-log). They will give you a clue about why your join is failing.

**Comment by clokep** [comment]
It looks like the various bits of this have been implemented. All the milestones are empty. I'm sure there's *more* that could be done, but overall this seems good to file as separate, specific follow-ups.
